src/libServer/WebsocketServer.cpp

```diff
@@ -290,8 +349,56 @@ void WebsocketServer::PrepareTxBlockAndTxHashes(
   m_jsonTxnBlockNTxnHashes["TxHashes"] = json_txhashes;
 }
 
+Json::Value CreateReturnAddressJson(const TransactionWithReceipt& twr) {
+  Json::Value _json;
+
+  _json["toAddr"] = twr.GetTransaction().GetToAddr().hex();
+  _json["fromAddr"] = twr.GetTransaction().GetSenderAddr().hex();
+  _json["amount"] = twr.GetTransaction().GetAmount().str();
+  _json["success"] = twr.GetTransactionReceipt().GetJsonValue()["success"];
+  _json["ID"] = twr.GetTransaction().GetTranID().hex();
+
+  return _json;
+}
+
+void WebsocketServer::ParseTxnLog(const TransactionWithReceipt& twr) {
+  LOG_MARKER();
+
+  const auto& txn_to_addr = twr.GetTransaction().GetToAddr();
+
+  const auto txn_from_addr = twr.GetTransaction().GetSenderAddr();
+
+  lock_guard<mutex> g(m_mutexTxnLogAddrHdlTracker);
+
+  auto addr_confirmed = txn_to_addr;
+
+  auto find_addr = m_txnLogAddrHdlTracker.m_addr_hdl_map.find(txn_to_addr);
+
+  if (find_addr == m_txnLogAddrHdlTracker.m_addr_hdl_map.end()) {
+    find_addr = m_txnLogAddrHdlTracker.m_addr_hdl_map.find(txn_from_addr);
+
+    if (find_addr == m_txnLogAddrHdlTracker.m_addr_hdl_map.end()) {
+      return;
+    }
+    addr_confirmed = txn_from_addr;
+  }
+  const auto log_json = CreateReturnAddressJson(twr);
+
+  lock_guard<mutex> h(m_mutexTxnLogDataBuffer);
+  for (const connection_hdl& hdl : find_addr->second) {
+    m_txnLogDataBuffer[hdl][addr_confirmed].append(log_json);
+  }
+}
+
+void WebsocketServer::ParseTxn(const TransactionWithReceipt& twr) {
```

Comment:
I'm concerned about the performance here, because each transaction is processed twice and it looks like both ParseTxnxxx functions have common code blocks (including mutex acquisition and map searches). Can you merge the two functions instead to make this more efficient?

Reply:
The mutex acquistions and search variables are different, if you look closely.